### PR TITLE
Store test data with report

### LINKED ISSUE #65: Store and report results for `test trace`
This is the follow-up issue for #27 (and PR #61), which implement the basic functionality of producing and executing Pytest tests from traces.

The task of this issue is, during the execution or upon finishing of testing the current trace, to retrieve and store in `reports` folder, under the name that corresponds to the test being executed:
 - full transaction data for each submitted transaction
 - results of executing the transaction
 - node logs
 - full Pytest output

The user might be given an option to either store the above results for all, or only for failed tests. This is a nice-to-have, and not required.

### PR DESCRIPTION
**Pull request by rnbguy**
Closes #65

### CODE REVIEW

File: atomkraft/test/__init__.py [hunk 1]
```diff
@@ -27,6 +27,9 @@ def RequiredFileOption(help, default):
 def trace(
     # currently, require the trace to be present.
     # later, there will be an option to pick up the last one from the model
+    verbose: bool = typer.Option(
+        False, "--verbose", "-v", help="Output logging on console"
+    ),
```
Comment (rnbguy) [inline atomkraft/test/__init__.py:32]: A `--verbose` flag to print output only when it's asked.

File: atomkraft/test/trace.py [hunk 2]
```diff
@@ -55,4 +56,29 @@ def test_trace(trace: PathLike, reactor: PathLike, keypath: str):
             )
         )
     print(f"Executing {test_name} ...")
-    pytest.main(["-s", test_path])
+
+    report_dir = root / "reports" / test_name
+    report_dir.mkdir(parents=True)
+
+    logging_file = report_dir / "log.txt"
```
Comment (rnbguy) [inline atomkraft/test/trace.py:73]: `logging_file` is where `logging` data (e.g. `logging.info("something happened")`) is stored.

File: atomkraft/test/trace.py [hunk 3]
```diff
@@ -55,4 +56,29 @@ def test_trace(trace: PathLike, reactor: PathLike, keypath: str):
...
+
+    pytest_report_file = report_dir / "report.jsonl"
```
Comment (rnbguy) [inline atomkraft/test/trace.py:75]: Everything else (e.g. `stdout`, `stderr`, pytest report) is recorded in `pytest_report_file`.

Comment (rnbguy) [reply atomkraft/test/trace.py:75]: `logging` data is also captured here. But `pytest_report_file` is JSON Lines file. One has to process it with JSON processor to filter the outputs. `logging_file` stores `logging` data in simple text.

File: atomkraft/test/trace.py [hunk 4]
```diff
@@ -55,4 +56,29 @@ def test_trace(trace: PathLike, reactor: PathLike, keypath: str):
...
+    if verbose:
+        pytest_args.append("-rP")
+        pytest_args.append("--log-cli-level=INFO")
+
+    pytest.main(pytest_args + [test_path])
+
+    default_nodes_dir = root / ".atomkraft" / "nodes"
+    report_nodes_dir = report_dir / "nodes"
+
+    shutil.copytree(default_nodes_dir, report_nodes_dir)
```
Comment (rnbguy) [inline atomkraft/test/trace.py]: Copy node data from `.atomkraft` directory to `report` directory.

File: atomkraft/test/__init__.py [hunk 5]
```diff
@@ -39,7 +42,9 @@ def trace(
     Test blockchain by running one trace
     """
 
-    test_trace(trace.name, reactor if reactor is None else reactor.name, keypath)
+    test_trace(
+        trace.name, reactor if reactor is None else reactor.name, keypath, verbose
```
Comment (hvanz) [inline atomkraft/test/__init__.py:46]: `reactor if reactor is None else reactor.name` 
Is this correct? This is equivalent to `None if reactor is None else reactor.name`

File: atomkraft/test/trace.py [hunk 6]
```diff
@@ -55,4 +56,29 @@ def test_trace(trace: PathLike, reactor: PathLike, keypath: str):
             )
         )
     print(f"Executing {test_name} ...")
-    pytest.main(["-s", test_path])
+
+    report_dir = root / "reports" / test_name
+    report_dir.mkdir(parents=True)
```
Comment (hvanz) [inline atomkraft/test/trace.py:71]: What happens if the directory already exists?

Comment (hvanz) [reply atomkraft/test/trace.py:71]: Now I see that `test_name` has a timestamp.

File: atomkraft/test/trace.py (hunk 4 above)
Comment (hvanz) [reply atomkraft/test/trace.py]: Probably put all this code in a separate function?

File: atomkraft/test/trace.py [hunk 7]
```diff
@@ -55,4 +56,29 @@ def test_trace(trace: PathLike, reactor: PathLike, keypath: str):
...
+    pytest_report_file = report_dir / "report.jsonl"
+
+    pytest_args = [
+        "--log-file-level=INFO",
+        f"--log-file={logging_file}",
+        f"--report-log={pytest_report_file}",
+    ]
+
+    if verbose:
+        pytest_args.append("-rP")
```
Comment (hvanz) [inline atomkraft/test/trace.py:84]: I tried to generate a report for the transfer example. The jsonl file and the nodes directory are ok, but log.txt is empty. Maybe something to do with the arguments to pytest?

**Review by hvanz — COMMENTED** [review]
I run `test trace` in the transfer example and it generates the expected files, except that log.txt is empty. Also I have a suggestion for moving the report-related code to a separate function.

File: atomkraft/test/trace.py (hunk 7 above)
Comment (rnbguy) [reply atomkraft/test/trace.py:84]: `log.txt` would be empty if there was no logging output. Currently, none of our examples uses `logging` module.

I can avoid copying the file, if the file is empty.

File: atomkraft/test/__init__.py (hunk 5 above)
Comment (rnbguy) [reply atomkraft/test/__init__.py:46]: I did not write this line. I had to modify it for the call to add the `verbose` argument. But you are right. I will take a look.

Comment (rnbguy) [reply atomkraft/test/__init__.py:46]: It is fine. It avoids accessing `reactor.name` if `reactor` is `None`.

File: atomkraft/test/trace.py (hunk 7 above)
Comment (rnbguy) [reply atomkraft/test/trace.py:84]: For now, let's leave it like this. Because this file is created by `pytest` directly.

**Comment by hvanz** [comment]
Looks good. Go ahead and merge when you are ready.

**Pull request merged**